### DISCUSSION TITLE
Confused about the Video Pipeline

### DISCUSSION THREAD

**Discussion by NicholasFlamy** [blakeblackshear/frigate#9682] [Ask A Question]
https://docs.frigate.video/frigate/video_pipeline#detailed-view-of-the-video-pipeline

The diagram makes it look like the detect stream is recorded in the recording store unless you have 24/7 recording. From what I understand, the detect stream is processed and if it makes an event it saves the record stream that matches the timestamps. If for example the snapshots are taken from the detect stream but event clips are saved from the record stream, then I think either the diagram should be updated with stars or something, or there should be some written clarification (if someone can clarify this I'm willing to explain it the best I can and PR the docs)

I see that the birdseye ending only has arrows leading from the detect stream. I assume that means it uses the detect stream?
The snapshot store only has an arrow from the object detection so that leads me to believe it's from the detect stream?

**Comment by NicholasFlamy** [comment]
So I just did a test. The snapshots come from the detect stream (as of 0.12.1) and the footage for the event is from the record stream. The Birdseye uses the detect stream. It could use some clarification. Also, why does the birdseye use the detect stream?

**Comment by NickM-27** [comment]
> The diagram makes it look like the detect stream is recorded in the recording store unless you have 24/7 recording

not sure I see that, the detect stream does not connect to the recordings store directly in any way, only the "detection" is added to the recording store. Perhaps this could be clarified that this refers to data not camera stream

> Also, why does the birdseye use the detect stream?

because birdseye is a composite of multiple streams. To do this the stream must be decoded, so using the high res stream would require a completely separate decode process and that would use a lot more CPU resources for very little gain in quality

**Reply by NicholasFlamy** [reply]
> Perhaps this could be clarified that this refers to data not camera stream

This is exactly what I wanted clarified, it doesn't specify that it's only gathering the data and timestamps to mark where in the recorded stream the event is, which is why I wasn't sure.

> because birdseye is a composite of multiple streams. To do this the stream must be decoded, so using the high res stream would require a completely separate decode process and that would use a lot more CPU resources for very little gain in quality

Huh, okay, I guess that makes sense. I guess I didn't understand how Birdseye worked, I kinda thought it was just pulling the streams, but that doesn't make sense since it can be restreamed.

**Reply by NicholasFlamy** [reply]
> not sure I see that, the detect stream does not connect to the recordings store directly in any way, only the "detection" is added to the recording store.

So I'm attaching a screenshot.
The red circles aren't clear about which one is footage that is saved vs which one is just information about the footage, maybe the object detection and motion detection saving could go to the database (new entry into the visualization) rather than the recording store to make it clearer that it specifies the events but is not the detect stream footage? And the "video segments (retain all)" could be changed to make sense in this context?
The blue circles make more sense since one specifies the decoded frames and the others just say motion event or object event which in contrast with "decoded frames" sounds more like the data rather than the footage.
The yellow circle is obvious if the rest makes sense because it's the only one going to snapshots.

![Screenshot 2024-02-05 164801](https://github.com/blakeblackshear/frigate/assets/30300649/c16c653c-9648-4f62-834a-5c907eb74646)

**Comment by NicholasFlamy** [comment]
One more question: How resource intensive is setting the detect resolution lower than native? I understand that with lowering the FPS (like 15 from source to 5 in detect config) it is just dropping frames with no load from doing that (like picking every 3rd frame.)

I'm wondering, if I went from 4K to 1080p, would it sample 4 pixels to create one (resource intensive), or would it nearest neighbor and pick one of 4 pixels and use that (very easy on resources)?

**Reply by NickM-27** [reply]
the only time it is recommended to do downscaling from the main stream is if you have a GPU and are using ffmpeg presets, that will use the GPU to do the scaling
